### PR TITLE
feat(SidebarE): Implement new experimental Sidebar

### LINKED ISSUE #965: Implement new experimental Sidebar
In its current form, the sidebar is lacking in functionality and overall usefulness. This iteration aims to provide DS users with a modular and adaptable sidebar.

Screenshot 2020-05-15 at 12 24 57

### Why v3?
In order for the new sidebar capabilities to work, we have to break a few prop APIs and layouts. This will require effort for the DS user to implement, so this improvement is scheduled for the next breaking release (`v3`). The main breaking changes that are needed are:

- A wrapper to surround the main application, allowing the main app content to resize when the sidebar is expanded and contracted
- More useful props (e.g. `user` to apply the full name of the user to the sidebar, instead of just initials)
- Revamped Sidebar nav items, allowing for contextual menus

### PR DESCRIPTION
**Pull request by m7kvqbe1**
## Related issue

Closes #965

## Overview

Implement new experimental Sidebar.

## Reason

>In its current form, the sidebar is lacking in functionality and overall usefulness. This iteration aims to provide DS users with a modular and adaptable sidebar.

## Work carried out

- [x] Expanded user profile links
- [x] Collapsed menu item tooltip
- [x] Sign out / view profile hover menu
- [x] Layout companion component
- [x] Notifications
- [x] Expanded sub-menu items
- [x] Increase size of main sidebar
- [x] A11y enhancements
- [x] Add automated tests
- [x] Add transition animations

## Developer notes

- Initial implementation does not include the 'Search bar' feature.

- I haven't yet implemented any animations or transitions - requirements for these need to be supplied by designers.

- Additional PRs coming soon to change the styling of `Popover` and `Sheet` components.

- Export is currently commented out until we have full design and developer documentation.

- I've created another ticket for adding Design and Develop documentation.

### CODE REVIEW

File: packages/react-component-library/src/components/Tooltip/Tooltip.tsx [hunk 1]
```diff
@@ -6,7 +6,7 @@ import { PositionType } from '../../common/Position'
 import { TOOLTIP_POSITION } from '.'
 
 export interface TooltipProps extends PositionType {
-  children: React.ReactNode
+  children?: React.ReactNode
```
Comment (thyhjwb6) [inline packages/react-component-library/src/components/Tooltip/Tooltip.tsx]: I am wondering if we need a Storybook example or test to support this change.

File: packages/react-component-library/src/components/TopLevelNavigation/SidebarE/SidebarUser.tsx [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+      />
+    )}
+    placement={SHEET_PLACEMENT.RIGHT_BOTTOM}
+    width={SHEET_WIDTH}
+  >
+    <ol>{children}</ol>
+  </StyledSheet>
+)
+
+export const SidebarUser: React.FC<SidebarUserProps> = ({
```
Comment (thyhjwb6) [inline packages/react-component-library/src/components/TopLevelNavigation/SidebarE/SidebarUser.tsx:133]: All tests still pass when this function is changed to:

```
export const SidebarUser: React.FC<SidebarUserProps> = ({
  initials,
  link,
  name,
  children,
}) => {
  const { isOpen } = useContext(SidebarContext)

  if (isOpen) {
    return null
  }

  if (!isOpen && children) {
    return (
      <StyledSidebarUser data-testid="sidebar-user">
        <SidebarAvatarWithItems initials={initials}>
          {children}
        </SidebarAvatarWithItems>
      </StyledSidebarUser>
    )
  }

  return null
}
```

File: packages/react-component-library/src/components/TopLevelNavigation/SidebarE/SidebarWrapper.tsx [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import styled from 'styled-components'
+
+import { ComponentWithClass } from '../../../common/ComponentWithClass'
+
+const StyledSidebarWrapper = styled.div`
+  display: flex;
+  flex-direction: row;
+`
+
+export const SidebarWrapper: React.FC<ComponentWithClass> = ({
```
Comment (thyhjwb6) [inline packages/react-component-library/src/components/TopLevelNavigation/SidebarE/SidebarWrapper.tsx:11]: All tests still pass when this is:

```
export const SidebarWrapper: React.FC<ComponentWithClass> = ({
  className,
  children,
}) => {
  return null
}
```

File: packages/react-component-library/src/components/TopLevelNavigation/SidebarE/SidebarNavItem.tsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    ),
+    onClick,
+    'aria-label': linkElement.props.children,
+    'data-testid': 'sidebar-nav-item',
+  })
+
+  return (
+    <StyledSidebarNavItem
+      isActive={isActive}
+      onMouseEnter={(_) => setHasMouseOver(true)}
```
Comment (thyhjwb6) [inline packages/react-component-library/src/components/TopLevelNavigation/SidebarE/SidebarNavItem.tsx:146]: All tests pass when the setting of `onMouseEnter` and `onMouseLeave` are removed.

**Review by thyhjwb6 — COMMENTED** [review]
The code looks 👍 

There are a few areas which would benefit from more tests.

**Review by thyhjwb6 — APPROVED** [review]
👍

**Comment by hxltrhuxze** [comment]
Had a look through the PR and it looks great so far. Couple of feedback points from myself:

#### Named CSS Classes
There's a lot of styling such as:
```
// SidebarUser.tsx

const StyledSidebarUserText = styled.div`
  ...

  > div {
    ...

    span:first-of-type {
      ...
    }

    span:last-of-type {
      ...
    }
  }
`
```
What is our approach for css classes whilst using Styled Components? Selectors such as `> div span:first-of-type` can be brittle as it couples generic markup with specific components. Any modifications/updates to this component have to be careful as generic, unclassed HTML selectors can have knock-on effects to child components due to the cascade.

Ideally, we'd use named classes with a syntax closer to the following:
```
const StyledSidebarUserText = styled.div`
  ...
  // span:first-of-type
  .foo {
    ...
  }
`
```

#### Sidebar expansion button

Screenshot 2020-10-26 at 14 04 52

Can we have this button only appear on `mouseEnter` of the sidebar?

#### icon size
Screenshot 2020-10-26 at 14 06 11
Can we bump these to `18px` each?

## Animations

**On expansion**: The sidebar should animate outwards, with the text appearing after the sidebar has fully expanded, in order to prevent clipping of the labels.
```
- sidebar width expand
- fade in text
```


**On Collapse**: The collapse animation should be the reverse of the expansion.
```
- fade out text
- sidebar width collapse
```
Screenshot 2020-10-26 at 14 09 53

We should aim to achieve all the animations in under `0.3s`.

EDIT: Accidentally left the "Application Name" label in the screenshot - this should act in the same way as the nav links.

File: packages/react-component-library/src/components/TopLevelNavigation/SidebarE/SidebarUser.tsx [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+  | SidebarUserWithLinkProps
+  | SidebarUserWithItemsProps
+
+const SHEET_WIDTH = 106
+
+const { spacing, color, fontSize } = selectors
+
+const StyledSidebarUser = styled.div`
+  display: flex;
+  align-items; center;
```
Comment (hxltrhuxze) [inline packages/react-component-library/src/components/TopLevelNavigation/SidebarE/SidebarUser.tsx]:
```suggestion
  align-items: center;
```

**Pull request merged**